File: Solutions/46. Permutations/46. Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def perm(nums:List[int],res:List[List[int]],begin:int):
            if begin==len(nums):
                res.append(nums[:])
                return
            for i in range(begin,len(nums)):
                nums[i],nums[begin]=nums[begin],nums[i]
                perm(nums,res,begin+1)
                nums[i],nums[begin]=nums[begin],nums[i]
            return
        res=[]
        perm(nums,res,0)
        return res
                

File: Solutions/46. Permutations/46. Permutations.go
func permute(nums []int) [][]int {
    var res [][]int
    perm(nums,&res,0)
    return res
}
func perm(nums []int,res *[][]int,begin int){
    if len(nums)==begin{
        tmp:=make([]int,len(nums))
        copy(tmp,nums)
        *res=append(*res,tmp)
        return 
    }
    for i:=begin;i<len(nums);i+=1{
        nums[begin],nums[i]=nums[i],nums[begin]
        perm(nums,res,begin+1)
        nums[begin],nums[i]=nums[i],nums[begin]
    }
    return
}